feat(winbar): add configurable control buttons

noirbizarre opened the pull request:
Hi 👋🏼 

This PR adds an opt-in control bar widget.
By default, it is disabled, as this is a mouse-only widget, and it can be enabled with `winbar.controls.enabled=true`.

The buttons are dynamic, they are only actives when it's relevant (active debug session or not, session paused on a breakpoint or not...)  and they are greyed out when it's not.

It can be positioned at the left or the right (default) of the tabs with `winbar.controls.position="left"|"right"`. Positioned to the right, the widget is right aligned, and positioned to the left, the tabs are then right aligned.

**`winbar.controls.position="right"` (default)**

![image](https://github.com/user-attachments/assets/55982377-4d6d-42c3-8453-240a1395794b)

**`winbar.controls.position="left"`**

![image](https://github.com/user-attachments/assets/70df1f6a-89c7-4711-830f-b36c9c58b6cc)

Each icon can be changed with the `winbar.controls.icons` settings. It defaults to the `codicons` dedicated `debug` glyphs for each.

You can choose to display the buttons you want and their order with the `winbar.controls.buttons` table, default is `{"play", "step_into", "step_over", "step_out", "step_back", "run_last", "terminate", "disconnect"}`.

Each color can also be customized using the dedicated highlights

`terminate` and `disconnect` button can be hybrid with respectively `winbar.controls.terminate_run_last` and `winbar.controls.disconnect_run_last` (both default to `false`). When enabled, instead of being deactivated when no debug session is running, they are transformed into a `run_last` button, allowing to get rid of the `run_last` button if you want (some may still want it as it allows restarting a debug session when running).

As an example, here's my current setup:
```lua
winbar = {
	controls = {
		enabled = true,
        position = "left",
        buttons = {
            "play",
            "step_into",
            "step_over",
            "step_out",
            "terminate",
            "disconnect",
        },
        icons = {
            play = "",
            terminate = "",
        },
        terminate_run_last = true,
},
```
![image](https://github.com/user-attachments/assets/5e19949c-a861-4b95-ae94-0049c03c965d)
![image](https://github.com/user-attachments/assets/667a60a5-0a0c-4e9e-9767-1ecb83e27d20)

>[!NOTE]
> I did not added any screenshot to the README as I am not using thte same color scheme and the result would be inconsistent

igorlfs commented on lua/dap-view/events.lua:
```diff
@@ -9,9 +9,17 @@ local exceptions = require("dap-view.exceptions.view")
 local SUBSCRIPTION_ID = "dap-view"
 
+local refresh_winbar = function()
+    --- Refresh the winbar if the controls are enabled
```
No need for this comment

igorlfs replied:
Can we check if the controls are active inside `winbar.refresh_winbar`? As to not need another function

noirbizarre replied:
I tried, but no, we can't call directly `winbar.refresh_winbar` (now merged into `winbar.update_winbar()`) as events parameter are passed as first argument.
So I kept this one, but merged the 2 functions in `winbar` module.
The alternative is the opposite: remove this `refresh_winbar()` but keep 2 separate functions in the module.
I am fine with both approach, I just had to chose one, but I can move the function in the `winbar` module.

igorlfs commented on lua/dap-view/events.lua:
```diff
@@ -121,4 +131,22 @@ dap.listeners.after.event_terminated[SUBSCRIPTION_ID] = function()
+--- Refresh winbar on dap session state change events not having a dedicated event handler
+local events = {
+    "continue",
+    "disconnect",
+    "disconnect",
```
It looks like you really disconnecting :P

igorlfs commented on lua/dap-view/events.lua:
```diff
@@ -121,4 +131,22 @@ dap.listeners.after.event_terminated[SUBSCRIPTION_ID] = function()
+    "continue",
+    "disconnect",
+    "disconnect",
+    "event_continued",
```
Are both `continue` and `event_continued` necessary?

igorlfs commented on lua/dap-view/config.lua:
```diff
@@ -15,6 +16,25 @@ local M = {}
+---@field position 'left' | 'right'
+---@field buttons ControlType[] Buttons to show in the controls section.
+---@field icons ControlsIcons Icons for each button.
+---@field disconnect_run_last boolean Disconnect buttons becomes Run last without session.
+---@field terminate_run_last boolean Terminate buttons becomes Run last without session.
```
For now, I think we should leave these options for a followup PR. I don't like the idea of adding edge case options like that. What if someone comes with a request to override another button?

It's best to either have a more general fallback mechanism or to not have options at all (and instead opt for a more consistent design), either way is fine by me.

noirbizarre replied:
Actually I made this just to be able to submit a pull-request. My initial implementation was a single hybrid button for `restart`, `terminate` and `disconnect` but that was very opiniated.
Given I want to reduce the number of buttons, but still provide all buttons in the PR, I suggested another approach 👇🏼

igorlfs commented on README.md:
```diff
@@ -302,6 +334,18 @@ NvimDapViewLineNumber
 NvimDapViewSeparator
 NvimDapViewThread
 NvimDapViewThreadStopped
+NvimDapViewTab
+NvimDapViewTabSelected
+NvimDapViewControlNC
+NvimDapViewControlPlay
+NvimDapViewControlPause
+NvimDapViewControlStepInto
+NvimDapViewControlStepOut
+NvimDapViewControlStepOver
+NvimDapViewControlStepBack
+NvimDapViewControlRunLast
+NvimDapViewControlTerminate
+NvimDapViewControlDisconnect
```
This code block is getting big! Do you mind wrapping it in a `<details>` tag? It also makes sense to add the default links (and use a table instead), but I can do that later.

noirbizarre replied:
Done

igorlfs commented on README.md:
```diff
@@ -122,6 +122,13 @@ The console's default size (height) is resized to match your `nvim-dap-view` con
 ![console](https://github.com/user-attachments/assets/0980962c-e3da-4f16-af4c-786ef7fa4b18)
 
+You can also enable the controlbar which exposes some clickable buttons by settings `winbar.controls.enable`.
+The position can be configured with `winbar.controls.position` and the enabled buttons with `winbar.controls.buttons`. 
+The icons can be configured with `winbar.controls.icons`. 
```
Instead of listing (almost) all options available, we could be more brief. Something along the lines of "The control bar is fully customizable, checkout the options on default configuration section".

igorlfs commented on README.md:
```diff
@@ -122,6 +122,13 @@ The console's default size (height) is resized to match your `nvim-dap-view` con
 
 ![console](https://github.com/user-attachments/assets/0980962c-e3da-4f16-af4c-786ef7fa4b18)
 
+You can also enable the controlbar which exposes some clickable buttons by settings `winbar.controls.enable`.
```
```suggestion
You can also enable the control bar which exposes some clickable buttons by settings `winbar.controls.enable`.
```

igorlfs commented on lua/dap-view/options/winbar.lua:
```diff
@@ -142,4 +290,8 @@ M.update_winbar = function(section_name)
     end
 end
 
+M.refresh_winbar = function()
+    M.update_winbar(state.current_section)
+end
```
We don't need this function at all, actually. But maybe `update_winbar` should be called `refresh_winbar`?

noirbizarre replied:
I kept the name `update_winbar` but made the current_section` optional, so there is only one function

igorlfs commented on lua/dap-view/options/winbar.lua:
```diff
@@ -80,6 +81,128 @@ local winbar_info = {
+            dap.run_last()
+        end,
+    },
+    terminate = {
+        render = function()
+            local session = dap.session()
+            local config = setup.config.winbar.controls
+            local off_hl = config.terminate_run_last and "ControlRunLast" or "ControlNC"
+            local hl = session and "ControlTerminate" or off_hl
+            local off_icon = config.terminate_run_last and config.icons.run_last or config.icons.terminate
+            local icon = session and config.icons.terminate or off_icon
+            return _hl(icon, hl)
+        end,
+        action = function()
+            if setup.config.winbar.controls.terminate_run_last and not dap.session() then
+                dap.run_last()
+            else
+                dap.terminate()
+            end
+        end,
+    },
+    disconnect = {
+        render = function()
+            local session = dap.session()
+            local config = setup.config.winbar.controls
+            local off_hl = config.disconnect_run_last and "ControlRunLast" or "ControlNC"
+            local hl = session and "ControlDisconnect" or off_hl
+            local off_icon = config.disconnect_run_last and config.icons.run_last or config.icons.disconnect
+            local icon = session and config.icons.disconnect or off_icon
+            return _hl(icon, hl)
+        end,
+        action = function()
+            if setup.config.winbar.controls.disconnect_run_last and not dap.session() then
+                dap.run_last()
+            else
+                dap.disconnect()
+            end
+        end,
+    },
+}
```
This file is getting ginormous. Can you move this bad boy to another file?

noirbizarre replied:
Done, moved into its own module

noirbizarre commented:
I applied all reviews comments and opted for another approach.
I removed the hybrid buttons and added support for custom buttons which in fine is far more flexible.
It allows user to simply declare their own custom buttons like this:
```lua
winbar = {
  controls = {
    enabled = true,
    buttons = {
      "play",
      "step_into",
      "step_over",
      "step_out",
      "term_restart",
      "fun",
    },
    custom_buttons = {
      fun = {
        render = function() return "🎉" end,
        action = function() vim.print("🎊") end,
      },
      -- Stop/Restart button
      -- Double clic, middle clic or clic with a modifier disconnect
      term_restart = {
        render = function()
          local session = require("dap").session()
          local group = session and "ControlTerminate" or "ControlRunLast"
          local icon = session and "" or ""
          return "%#NvimDapView" .. group .. "#" .. icon .. "%*"
        end,
        action = function(clicks, button, modifiers)
          local dap = require("dap")
          local alt = clicks > 1 or button ~= "l" or modifiers:gsub(" ", "") ~= ""
          if not dap.session() then
            dap.run_last()
          elseif alt then
            dap.disconnect()
          else
            dap.terminate()
          end
        end,
      },
    },
  },
}
```
I think it opens way more possibilities and each one is free to do as he want.

WDYT ?

![image](https://github.com/user-attachments/assets/e4c939af-4d93-4223-b25f-655cf41db2d1)

igorlfs commented on lua/dap-view/highlight.lua:
```diff
@@ -17,10 +19,39 @@ local define_base_links = function()
+---@param text string
+---@param handler string
+---@param idx integer
+M.clickable = function(text, module, handler, idx)
```
I don't think we need this function. Not using it is a bit awkward, but it's only used twice.

noirbizarre replied:
Given I moved those helpers in a `statusline` submodule as suggested in a previous comment, I chose to kept it as it was barely readable without, formatting was making it even worse, and I think it gives semantics so readability and maintainability.

igorlfs commented on lua/dap-view/options/controls.lua:
```diff
@@ -0,0 +1,135 @@
+    return bar
+end
+
+---@return boolean
+M.enabled_left = function()
+    local config = setup.config.winbar.controls
+    return config.enabled and config.position == "left"
+end
+
+---@return boolean
+M.enabled_right = function()
+    local config = setup.config.winbar.controls
+    return config.enabled and config.position == "right"
+end
```
I think these are a bit overkill, they can be inlined

igorlfs reviewed (commented):
Thanks for the patience. We are getting there.

igorlfs commented:
> I think it opens way more possibilities and each one is free to do as he wants.
> 
> WDYT ?

That's a really nice approach! Make sure to add a section about that in the README. 

> My initial implementation was a single hybrid button for restart, terminate and disconnect but that was very opinionated.

That sounds like a nice example for the README.

noirbizarre commented on lua/dap-view/events.lua:
```diff
@@ -121,4 +130,20 @@ dap.listeners.after.event_terminated[SUBSCRIPTION_ID] = function()
+    "disconnect",
+    "event_exited",
+    "restart",
+    "terminate",
```
I tried them all individually to ensure I kept the minimum list possible.
Actually, `event_terminated` is required, or the buttons are not refreshed at the end of the run.
However, I have been able to remove the `terminate` handler

noirbizarre commented:
PR updated.
I applied all suggestions, implemented the suggested `winbar` refactoring for the refresh/update/redraw and added a README section for custom buttons with the example above folded (I put both the `fun` and `term_restart` buttons as I think it gives an overview of the possibility, from the most basic to a way a more complex case).
As I said, I did not added any screenshots to the README by consistency, given they would be using a different theme.

Note that doing this PR was both very interesting and fun! I learnt many things about NeoVim internals and DAP.

igorlfs commented on README.md:
```diff
@@ -335,6 +383,65 @@ vim.api.nvim_create_autocmd({ "FileType" }, {
 
 </details>
 
+### Custom buttons
```
Can you update the ToC to also include this topic? I'm frequently forgetting about it myself, I wonder if there's some way to automate it

noirbizarre replied:
Updated 👍🏼 

I think there might be some action to do that.
Maybe at some point, when the documentation grows, these kinds of small tasks become an argument for a dedicated documentation site (and IMHO, it could be a builtin GitHub feature)

igorlfs replied:
>  an argument for a dedicated documentation site

loved the idea!

igorlfs commented on README.md:
```diff
@@ -335,6 +383,65 @@ vim.api.nvim_create_autocmd({ "FileType" }, {
 
+### Custom buttons
+
+`nvim-dap-view` provides some default buttons, but you can also add your own. To do that, in the `controls` table you can use the `custom_buttons` table to declare your new button and then add it at the position you want in the `buttons` list. 
```
```suggestion
`nvim-dap-view` provides some default buttons for the control bar, but you can also add your own. To do that, in the `controls` table you can use the `custom_buttons` table to declare your new button and then add it at the position you want in the `buttons` list. 
```

igorlfs commented on README.md:
```diff
@@ -335,6 +383,65 @@ vim.api.nvim_create_autocmd({ "FileType" }, {
+
+An example adding 2 buttons:
+- `fun`: the most basic button possible, just prints "🎊" when clicked
+- `term_restart`: an hybrid button that acts as a stop/restart button. If the stop button is triggered by anything else than a click (middle click, right click, double click or click with a modifier), it will disconnect the session instead.
```
```suggestion
- `term_restart`: an hybrid button that acts as a stop/restart button. If the stop button is triggered by anything else than a single left click (middle click, right click, double click or click with a modifier), it will disconnect the session instead.
```

igorlfs commented on README.md:
```diff
@@ -335,6 +383,65 @@ vim.api.nvim_create_autocmd({ "FileType" }, {
+        action = function() vim.print("🎊") end,
+      },
+      -- Stop/Restart button
+      -- Double clic, middle clic or clic with a modifier disconnect instead of stop
```
```suggestion
      -- Double click, middle click or click with a modifier disconnect instead of stop
```

igorlfs commented on lua/dap-view/options/winbar.lua:
```diff
@@ -135,10 +144,16 @@ M.show_content = function(selected_section)
+    end
+end
+
+M.redraw_controls = function()
+    if setup.config.winbar.controls.enabled then
+        set_winbar_opt()
     end
 end
```
I believe we also have to check if the winbar is enabled (i.e., `show`) here?

noirbizarre replied:
Yes, we have to. Added 👍🏼

igorlfs commented on lua/dap-view/options/winbar.lua:
```diff
@@ -1,5 +1,8 @@
 local setup = require("dap-view.setup")
+local controls = require("dap-view.options.controls")
+local statusline = require("dap-view.util.statusline")
+local module = ...
```
This module line is truly some arcane magic. No need to worry about it though, it's just my first contact with this syntax

noirbizarre replied:
I know, this one is mind-blowing 🤯 I discovered it while doing this pull-request, it's new to me too!

igorlfs reviewed (approved):
Wonderful! Left some final comments

igorlfs commented:
> As I said, I did not added any screenshots to the README by consistency, given they would be using a different theme.

No worries! I'll add one after merging this PR, when rebasing in the watches branch.

> Note that doing this PR was both very interesting and fun! I learnt many things about NeoVim internals and DAP.

I'm glad you enjoyed! Overall, that's also my experience with the plugin as a whole.

noirbizarre commented:
There it is, the final version with the last comments applied! 🚀
